src/Illuminate/Foundation/Testing/TestCase.php

```diff
@@ -262,4 +267,115 @@ protected function callBeforeApplicationDestroyedCallbacks()
             }
         }
     }
+
+    /**
+     * This method is called when a test method did not execute successfully.
+     *
+     * @throws Throwable
+     */
+    protected function onNotSuccessfulTest(Throwable $exception): void
+    {
+        if (! $exception instanceof ExpectationFailedException || ! $this->latestResponse) {
+            parent::onNotSuccessfulTest($exception);
+        }
+
+        if ($lastException = $this->latestResponse->exceptions->last()) {
+            parent::onNotSuccessfulTest($this->appendExceptionToException($lastException, $exception));
+
+            return;
+        }
+
+        if ($this->latestResponse->baseResponse instanceof RedirectResponse) {
+            $session = $this->latestResponse->baseResponse->getSession();
+
+            if (! is_null($session) && $session->has('errors')) {
+                parent::onNotSuccessfulTest($this->appendErrorsToException($session->get('errors')->all(), $exception));
+
+                return;
+            }
+        }
+
+        if ($this->latestResponse->baseResponse->headers->get('Content-Type') === 'application/json') {
+            $testJson = new AssertableJsonString($this->latestResponse->getContent());
+
+            if (isset($testJson['errors'])) {
+                parent::onNotSuccessfulTest($this->appendErrorsToException($testJson->json(), $exception, true));
+
+                return;
+            }
+        }
+
+        parent::onNotSuccessfulTest($exception);
+    }
+
+    /**
+     * Append an exception to the message of another exception.
+     *
+     * @param  Throwable  $exceptionToAppend
+     * @param  Throwable  $exception
+     * @return Throwable
+     */
+    protected function appendExceptionToException($exceptionToAppend, $exception)
+    {
+        $exceptionMessage = $exceptionToAppend->getMessage();
+        $exceptionToAppend = (string) $exceptionToAppend;
+        $message = <<<"EOF"
+            The following exception occurred during the last request:
+
+            $exceptionToAppend
+
+            ----------------------------------------------------------------------------------
+
+            $exceptionMessage
+            EOF;
```

Comment:
I think we should put the `EOF` at the start of the line?

```suggestion
EOF;
```

Reply:
If we put it at the start of the line then we'll also need to move all of the other text inside it to the left as well, which I'm not a fan of.

I'm using the PHP 7.3+ [flexible heredoc syntax](https://laravel-news.com/flexible-heredoc-and-nowdoc-coming-to-php-7-3) here so the indentation starts from the closing tag (i.e. no indentation in this case)

Reply:
Looks like GitHub syntax highlighting might not be sure how to handle it though :woman_facepalming:

Reply:
Ah, gotcha!